server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageStats.java

```diff
@@ -0,0 +1,281 @@
+        checkByteSize(bytes);
+        getOrAdd(fieldName).proximityBytes += bytes;
+    }
+
+    public void addStoredField(String fieldName, long bytes) {
+        checkByteSize(bytes);
+        getOrAdd(fieldName).storedFieldBytes += bytes;
+    }
+
+    public void addDocValues(String fieldName, long bytes) {
+        checkByteSize(bytes);
+        getOrAdd(fieldName).docValuesBytes += bytes;
+    }
+
+    public void addPoints(String fieldName, long bytes) {
+        checkByteSize(bytes);
+        getOrAdd(fieldName).pointsBytes += bytes;
+    }
+
+    public void addNorms(String fieldName, long bytes) {
+        checkByteSize(bytes);
+        getOrAdd(fieldName).normsBytes += bytes;
+    }
+
+    public void addTermVectors(String fieldName, long bytes) {
+        checkByteSize(bytes);
+        getOrAdd(fieldName).termVectorsBytes += bytes;
+    }
+
+    public IndexDiskUsageStats add(IndexDiskUsageStats other) {
+        other.fields.forEach((k, v) -> getOrAdd(k).add(v));
+        this.indexSizeInBytes += other.indexSizeInBytes;
+        return this;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        final PerFieldDiskUsage total = total();
+        builder.field(INDEX_SIZE, new ByteSizeValue(indexSizeInBytes));
+        builder.field(INDEX_SIZE_IN_BYTES, indexSizeInBytes);
+
+        // all fields
+        builder.startObject("all_fields");
+        total.toXContent(builder, params);
+        builder.endObject();
+
+        // per field
+        builder.startObject("fields");
+        {
+            final List<String> sortedFields = fields.keySet().stream().sorted().collect(Collectors.toList());
+            for (String field : sortedFields) {
+                builder.startObject(field);
+                fields.get(field).toXContent(builder, params);
+                builder.endObject();
+            }
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this);
+    }
+
+    /**
+     * Disk usage stats for a single field
+     */
+    public static final class PerFieldDiskUsage implements ToXContentFragment, Writeable {
+        private long termsBytes;
+        private long proximityBytes;
+        private long storedFieldBytes;
+        private long docValuesBytes;
+        private long pointsBytes;
+        private long normsBytes;
+        private long termVectorsBytes;
+
+        private PerFieldDiskUsage() {
+
+        }
+
+        private PerFieldDiskUsage(StreamInput in) throws IOException {
+            termsBytes = in.readZLong();
+            proximityBytes = in.readZLong();
+            storedFieldBytes = in.readZLong();
+            docValuesBytes = in.readZLong();
+            pointsBytes = in.readZLong();
+            normsBytes = in.readZLong();
+            termVectorsBytes = in.readZLong();
```

Comment:
why zlongs, can these numbers even be negative?

Reply:
Yes, fixed in https://github.com/elastic/elasticsearch/pull/74051/commits/fc67c3cba38f497d91e24ff8b00038900385e743